crates/router/src/services/authentication/detached.rs

```diff
@@ -61,13 +64,15 @@ impl ExtractedPayload {
                 message: format!("`{}` header not present", HEADER_AUTH_TYPE),
             })?;
 
+        let key_id = headers
+            .get(HEADER_KEY_ID)
+            .and_then(|value| value.to_str().ok())
+            .and_then(|key_id| ApiKeyId::try_from(Cow::from(key_id.to_string())).ok());
```

Comment:
```suggestion
        let key_id = headers
            .get(HEADER_KEY_ID)
            .and_then(|value| value.to_str().ok())
            .map(|key_id| ApiKeyId::try_from(Cow::from(key_id.to_string())))
            .transpose()
            .change_context(ApiErrorResponse::InvalidRequestData {
                 message: format!("`{}` header is invalid or not present", HEADER_KEY_ID),
             })?;

```

Reply:
So this does a few transformations.
```
 get :: a -> Option<b>
 and_then :: b -> Option<c>
 and_then :: c -> Option<Result<d, e>>
 transpose :: Option<Result<d, e>> -> Result<Option<d>, e>
 change_context :: Result<Option<d>, e> -> E -> Result<Option<d>, E>
 ? :: Result<Option<d>, E> -> Option<d>
 ```

Basically, we are only marking this as option, if
1. the header is not present
2. the value can't be converted into string
but not when
1. it fails to be parsed as a valid `key_id`

This makes sure that the key_id is valid, while preserving the behavior of the key id being optional

Reply:
Thanks for the explanation. Added required changes.